src/backend/metal/src/command.rs

```diff
@@ -1435,15 +1437,179 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
         U: IntoIterator,
         U::Item: Borrow<pso::ClearRect>,
     {
-        unimplemented!()
+        // gather vertices/polygons
+        let de = self.state.framebuffer_inner.extent;
+        let mut vertices = Vec::new();
+        println!("xx");
+        for rect in rects {
+            let r = rect.borrow();
+            for layer in r.layers.clone() {
+                let data = [
+                    [
+                        r.rect.x,
+                        r.rect.y,
+                    ],
+                    [
+                        r.rect.x,
+                        r.rect.y + r.rect.h,
+                    ],
+                    [
+                        r.rect.x + r.rect.w,
+                        r.rect.y + r.rect.h,
+                    ],
+                    [
+                        r.rect.x + r.rect.w,
+                        r.rect.y,
+                    ],
+                ];
+                // now use the hard-coded index array to add 6 vertices to the list
+                //TODO: could use instancing here
+                // - with triangle strips
+                // - with half of the data supplied per instance
+
+                for &index in &[0usize, 1, 2, 2, 3, 0] {
+                    let d = data[index];
+                    vertices.push(ClearVertex {
+                        pos: [
+                            d[0] as f32 / de.width as f32,
+                            d[1] as f32 / de.height as f32,
+                            0.0, //TODO: depth Z
+                            layer as f32,
+                        ],
+                    });
+                }
+            }
+        }
+
+        let mut commands = Vec::new();
+        let mut vertex_is_dirty = true;
+
+        //  issue a PSO+color switch and a draw for each requested clear
+        let mut pipes = self.shared.service_pipes
+            .lock()
+            .unwrap();
+
+        for clear in clears {
+            let key = match *clear.borrow() {
+                com::AttachmentClear::Color(index, value) => {
+                    let (format, channel) = self.state.framebuffer_inner.colors[index];
+                    //Note: technically we should be able to derive the Channel from the
+                    // `value` variant, but this is blocked by the portability that is
+                    // always passing the attachment clears as `ClearColor::Float` atm.
+                    let raw_value = com::ClearColorRaw::from(value);
+                    commands.push(soft::RenderCommand::BindBufferData {
+                        stage: pso::Stage::Fragment,
+                        index: 0,
+                        bytes: unsafe {
+                            slice::from_raw_parts(raw_value.float32.as_ptr() as *const u8, 16)
+                        }.to_owned(),
+                    });
+                    ClearKey {
+                        color: Some((format, index as u8, channel.unwrap())),
```

Comment:
What happens when `index > 1` here? Won't `let ps_name = format!("ps_clear{}_{}", index, s_channel);` fail to find the appropriate shader?

Reply:
Yep, it feels like this either needs a constant-based RT index selection that comes in Metal 2.0, or mamy many repeated shaders, or some sort of a generator for those (TODO).

Reply:
Ok, yeah, maybe a simple generator wouldn't be too hard